### PR TITLE
[PD-1943] Group activities in role creation by application

### PR DESCRIPTION
**Pull request by lemniscate613**
# Summary:

* Enable users to assign activities on a per app basis:

screen shot 2015-12-09 at 2 52 36 pm

* Transform activity names on front end only (from "create contact" to "contact - create"). Pertains to Activities, Roles, and Edit Role views. This groups activities visually better.

* Break out generalized functions in /gulp/src/util/ folder

* Start adding tests for all new /util/ functions

# Review:
```
cd /gulp
npm install
gulp
```
CRUD roles at http://127.0.0.1:8000/manage-users/#/roles

# But Wait There's More:

No there isn't. This is a "bitesize" PR.

abc_gma_billy_mays_090629_mn

### CODE REVIEW

File: gulp/src/manageusers/Role.jsx [hunk 1]
```diff
@@ -18,8 +23,10 @@ class Role extends React.Component {
       activitiesMultiselectHelp: '',
       roleName: '',
       roleNameHelp: '',
-      availableActivities: [],
-      assignedActivities: [],
+      availablePRMActivities: [],
```
Comment (lemniscate613) [inline gulp/src/manageusers/Role.jsx]: @darrint Okay it's official. Managing state is becoming cumbersome. Flux!!

**Comment by aspidites** [comment]
## Functionality
- Despite getting responses back, both lists are epty for me. I ran `npm install` in addition to `gulp`. My user is assigned the admin role
- Is it expected that I be able to manipulate myself as a user?

## Usability
- It's not apparently obvious to me that I have to click the section name in order to expand it
- Do we care that in order to review which permissions I've added, I have to expand a section, since you can only have one section expanded at a time?

## Screenshot
2015-12-09-155223_622x643_scrot

**Comment by aspidites** [comment]
Since you're updating the backend code, it'd be nice if there were tests ensuring that the shape of the response was what you wanted. That will also make it easier to refactor the backend code later.

**Comment by lemniscate613** [comment]
@aspidites API test are a good idea. Unless I misunderstand, they're handled by this task: https://directemployersfoundation.atlassian.net/browse/PD-1994

**Comment by koonsies** [comment]
I agree with @aspidites in regards to the accordions. Is there an "expand me" arrow that can be added? Activities loaded fine for me.

**Comment by koonsies** [comment]
Are we intending on adding a similar grouping to the Activities page?

**Comment by lemniscate613** [comment]
@koonsies Not a bad idea. I'll add that to a different task and see how Sole prioritizes it.

**Comment by aspidites** [comment]
@lemniscate613 re: tests:

1. With the way my last try went, it's clear I don't know what response your frontend is expecting to return. As such, I think this is better tackled by you, rather than me, as such, I've reassigned the ticket.
2. My intent with that ticket was that new development wouldn't start until the tests were done. My recent failure to complete the PEP-8 changes demonstrate the importance of unit tests.

File: gulp/src/util/spec/filterActivitiesForAppAccessSpec.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+      'app_access': 2,
+      'name': 'delete user',
+      'description': 'Remove existing users.',
+    },
+  },
+];
+
+describe('filterActivitiesForAppAccess', () => {
+  it('can filter out activities by app access', () => {
+    let filteredActivities = filterActivitiesForAppAccess(input, 1);
```
Comment (koonsies) [inline gulp/src/util/spec/filterActivitiesForAppAccessSpec.js]: A length check may be nice here. Expect 3 results for the first, 2 for the second.

**Comment by koonsies** [comment]
This isn't directly related to this ticket, but we should probably do something about the Role display screen. With only two roles, I need to scroll down to get to "Add Role". This is mostly due to all of the associated activities being listed for each role. Perhaps a different system could be devised, such as 

Activities (12/32)
.. or ..
Activities (12/32)
 - User Management (8/20)
 - PRM (4/12)

Or perhaps put an "Add Role" button at the top also.

Not sure if this is a priority, but something that may warrant some thought.

**Comment by koonsies** [comment]
Another thought.. Do we want to let the users modify the Admin role? This may have been discussed previously, but should the name/activities associated with Admin role be locked to prevent users from locking themselves out of things?

**Comment by koonsies** [comment]
Aside from my mostly unrelated questions :+1:

**Comment by aspidites** [comment]
@koonsies , conceptually, the Admin role should be read-only. Activities are added to it automatically and a user should not be permitted to add or remove activities from this role. Not that you asked, but a company can have more than one admin, but the interface should enforce the fact that at least one user should be assigned this role, as discussed during one of our meetings a few weeks back.

**Comment by koonsies** [comment]
@aspidites that was my thought as well. It felt a bit weird to be able to change the name of admin, let alone remove activities from the role.

**Comment by koonsies** [comment]
Where are we at with this PR? @aspidites did you get the API issues resolved that you had run into? Are unit tests coming in a future PR?

**Comment by aspidites** [comment]
With tests being written, I had no intention of merging this until after the tests were merged, as we could then use those tests to ensure that this PR is behaving as expected.

**Comment by lemniscate613** [comment]
@aspidites @koonsies Thanks for the feedback!

RE: tests, I paused work on this PR while creating tests. From now on, tests will be apart of the relevant PR. For now, they are in this PR and ready to be QC'd and merged: https://github.com/DirectEmployers/MyJobs/pull/1969 Once that's merged I will:

1. Pull quality-control int this branch
2. Finish the down arrow on the accordions
3. Address a few other bits from comments here

**Comment by aspidites** [comment]
If the accordion stuff is being addressed elsewhere, I'm :+1:  on this once @koonsies  and anyone else's feedback is fully addressed.

**Comment by lemniscate613** [comment]
> With only two roles, I need to scroll down to get to "Add Role". This is mostly due to all of the associated activities being listed for each role.

Good point @koonsies. This will be addressed by abbreviating the list of activities: https://directemployersfoundation.atlassian.net/browse/PD-1944

I created a Jira ticket ( https://directemployersfoundation.atlassian.net/browse/PD-2043 ) to get feedback on these points. If they are implemented they will be in a different PR:

* Maybe make the Admin role not editable (i.e. activites and role name)
* The frontend/API should enforce the fact that at least one user (per company) should be assigned the Admin role.

**Comment by lemniscate613** [comment]
I broke the accordion bit off into this Jira ticket (to be categorized by Jason S): https://directemployersfoundation.atlassian.net/browse/PD-2047

The protections against editing Admin role, etc. is handled in this Jira ticket: https://directemployersfoundation.atlassian.net/browse/PD-2046

**Pull request merged**